KAFKA-16012: Ensure new leader information merged correctly with the current metadata

philipnee opened the pull request:
A small bug in KIP-951 that when we perform metadata update during the leader change, we replace all topic metadata with the partial response from the fetch response, which leads to the consumer only fetching from a subset of partitions.

In this PR, the predicate we pass in to `MetadataCache#mergeWith` always return true, because we always want to retain the current topic metadata.

philipnee committed:
fix partial metadata update for KIP-951

clean up test

adding myank's comment

wip

Update ConsumerMetadataTest.java

philipnee commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -1071,18 +1071,25 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
     public void testUpdatePartitionLeadership() {
         Time time = new MockTime();
```
We probably need to refactor this test a certain point.  I tried to do it but as I moved something, we immediately exceeded the 160 line limit.

jolshan replied:
Maybe we should split up the test for each of the features (2 partitions, invalid topic, unauthorized topic, 1 internal topic). All the common logic can be in a helper?

philipnee replied:
sounds good - i was trying to limit the scope of change that's why i left this comment. i'll refactor this.

philipnee replied:
i did file this jira: https://issues.apache.org/jira/browse/KAFKA-16027

jolshan replied:
Sorry for confusion, I think you were right to limit the scope of the change. Let's do the refactor as a followup.

philipnee replied:
Alright - I'll revert some commits and re-push

philipnee committed:
clean up test

philipnee committed:
Update MetadataTest.java

Update MetadataTest.java

philipnee commented:
Here is the test results
```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.8.14
session_id:       2023-12-15--001
run time:         4 minutes 22.315 seconds
tests run:        4
passed:           4
flaky:            0
failed:           0
ignored:          0
================================================================================
test_id:    kafkatest.tests.core.group_mode_transactions_test.GroupModeTransactionsTest.test_transactions.failure_mode=clean_bounce.bounce_target=brokers
status:     PASS
run time:   1 minute 40.624 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.group_mode_transactions_test.GroupModeTransactionsTest.test_transactions.failure_mode=clean_bounce.bounce_target=clients
status:     PASS
run time:   2 minutes 5.353 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.group_mode_transactions_test.GroupModeTransactionsTest.test_transactions.failure_mode=hard_bounce.bounce_target=clients
status:     PASS
run time:   2 minutes 3.331 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.group_mode_transactions_test.GroupModeTransactionsTest.test_transactions.failure_mode=hard_bounce.bounce_target=brokers
status:     PASS
run time:   2 minutes 41.633 seconds
--------------------------------------------------------------------------------
```

philipnee committed:
Clean up

Update MetadataTest.java

clean up test

jolshan commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -427,10 +427,15 @@ public synchronized Set<TopicPartition> updatePartitionLeadership(Map<TopicParti
-                    updatePartitionMetadata, Collections.emptySet(), Collections.emptySet(),
-                    Collections.emptySet(), cache.cluster().controller(), topicIdsForUpdatedTopics,
-                    (topic, isInternal) -> !updatedTopics.contains(topic));
+        // Always merge the updated partitions with the existing topic metadata
```
nit: Can we maybe include a note about the desired behavior -- FETCH requests can include some partition changes, and when we keep existing topic metadata, we take the new partition changes but keep all partitions that weren't changed?

philipnee committed:
refacotr test

philipnee committed:
make the comment clearer

philipnee commented:
hi @jolshan - thank you so much for the review. I have addressed the comment in the recent commits.

jolshan commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -1068,143 +1067,328 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
+        assertEquals(1, metadata.fetch().partition(tp1).leader().id());
+
+        // "input-topic" partition 1 leader changes from node 1 to node 2
+        metadata.updatePartitionLeadership(
```
Is this the KIP-951 specific logic?

philipnee replied:
yes this was introduced in kip-951

jolshan replied:
Maybe we could just have a comment there (in the 1130 comment) 
something like: In KIP-951 we can update leadership info for a single partition. "input-topic partition 1...(everything else you said)

jolshan commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -1068,143 +1067,328 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
 
-        // Setup metadata with initial set of 2 partitions, 1 each across topics, with 5 nodes.
-        // Also setup, 1 invalid topic, 1 unauthorized topic, 1 internal topic.
+        String clusterId = "kakfa-cluster";
```
nit: kafka

jolshan commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -1068,143 +1067,328 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
+        Time time = new MockTime();
+
+        String clusterId = "kakfa-cluster";
+        String unauthroizedTopics = "unauthorized-topic";
```
nit: unauthorized and kafka

philipnee committed:
pr comments

philipnee committed:
revert the changes

revert

philipnee commented:
Hey @jolshan - Thanks again for the review.  Addressed your comments in the latest commits. The only client-related failure is `testExpandingTopicSubscriptions`, a known flaky test.  I will disable it in a different PR.

msn-tldr commented:
@philipnee lgtm, thanks for working on this.

jolshan commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -1067,25 +1067,102 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
+        PartitionMetadata part1Metadata = new PartitionMetadata(Errors.NONE, tp11, Optional.of(1), Optional.of(100), Arrays.asList(1, 2), Arrays.asList(1, 2), Arrays.asList(3));
         Uuid topic1Id = Uuid.randomUuid();
+        TopicPartition tp12 = new TopicPartition(topic1, 1);
+        PartitionMetadata part12Metadata = new PartitionMetadata(Errors.NONE, tp12, Optional.of(2), Optional.of(200), Arrays.asList(2, 3), Arrays.asList(2, 3), Arrays.asList(1));
```
nit: (we can address in a followup refactor) this naming is a bit confusing. Maybe all should be part11Metadata, part12Metadata, part21Metadata or even just tp10Metadata, tp11Metadata etc.

philipnee replied:
yap - i was trying to be consistent with the previous variable names such as `PartitionMetadata part1Metadata`. I'll follow up with the changes in a separated refactor PR.

jolshan commented on clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:
```diff
@@ -1185,9 +1263,14 @@ else if (partition.equals(internalPart))
-        List<Node> expectedNodes, String expectedClusterId, Set<String> expectedUnauthorisedTopics, Set<String> expectedInvalidTopics, Set<String> expectedInternalTopics,
-        Node expectedController, Map<String, Uuid> expectedTopicIds) {
+    void validateForUpdatePartitionLeadership(Metadata updatedMetadata,
+                                              PartitionMetadata part1Metadata, PartitionMetadata part2Metadata, PartitionMetadata part12Metadata,
```
nit: (for refactor) -- if we decided to put all the arguments on new lines, let's do it for the all the arguments. The ones with three are a bit harder to read.

philipnee replied:
thanks - would you mind leaving this for the refactor PR?

jolshan replied:
yup that's fine.

jolshan commented:
I was looking at some of the test failures and wasn't seeing the OffsetsApiIntegrationTests failing elsewhere. Can we check those? Running locally may not be enough so I will also restart the build.

philipnee commented:
hi @jolshan - it seems like `OffsetsApiIntegrationTests` is connect related test, in `org.apache.kafka.connect.integration`.  in the other PR, some of the tests in the integration test module also fail sometimes.

jolshan commented:
I'm seeing some OOMs and such again. I will run one more time. @philipnee if you don't mind just double checking locally for the tests on your end. I'm not seeing any failing consistently, but just want to be extra careful near the code freeze. I will check back in few hours or so.

philipnee commented:
Hey @jolshan - Thanks.  OOM also happens to another PR of mine see #15035 - i thought it was introduced by the AsyncKafkaConsumerTest but apparently there's more to it... Let me verify the tests locally and I'll the results there.

philipnee commented:
![image](https://github.com/apache/kafka/assets/1930388/6d79e6b4-76dc-49cc-8cbb-9b237a5dc7ef)

Unit tests seems fine, there is 1 known flaky test under AbstractCoordinator and 1 related to AsyncKafkaConsumer.  Both are unrelated to the metadata change here.

philipnee commented:
Consumer Integration tests also seems fine.
![image](https://github.com/apache/kafka/assets/1930388/df5307c2-0735-4789-9b87-e6d5eb9f1e7d)

jolshan commented:
Thanks for checking @philipnee. I will wait for this run to complete and we should be good to go.

philipnee commented:
Thank you!

philipnee commented:
jenkins doesn't seem to be in a great state.  there are a bunch of zkMigrationIntegrationTest and ReplicaManagerTest complaining timeout and `Found unexpected 1 NonDaemon threads` (there are others as well)

jolshan commented:
Yeah. I was looking at builds with @dajac for his PR too. The java 8 one here OOM'd. 
I think the ZkMigrationTests are causing the thread leak for the replica manager since I see replica manager tests failing with this issue on the same builds as the ZKMigrationTest failures and not a lot of commonalities otherwise.

I will file a JIRA for that. I will start one more time to get a super clear signal and will come back afterwards.

jolshan commented:
jira here: [KAFKA-16045](https://issues.apache.org/jira/browse/KAFKA-16045)

philipnee committed:
Merge branch 'apache:trunk' into kip-951-fix

philipnee commented:
Hey @jolshan seems like the build finished without oom, what do you think?

jolshan commented:
Ok -- looks good. I will merge and backport to 3.7.
